src/v/cluster/partition_leaders_table.cc

```diff
@@ -90,45 +94,67 @@ void partition_leaders_table::update_partition_leader(
+         * doing it only once for each topic.
+         **/
+        version topic_map_version_snapshot = _topic_map_version;
+        auto [t_it, _] = _topic_leaders.try_emplace(topic.tp_ns);
+
+        const bool is_controller = topic.tp_ns == model::controller_nt;
+        bool present_in_table = _topic_table.local().contains(topic.tp_ns);
+        auto last_applied_revision
+          = _topic_table.local().last_applied_revision();
+        for (auto& p : topic.partitions) {
+            if (!p.leader_id.has_value()) {
+                continue;
+            }
+
+            const auto topic_removed = p.revision_id <= last_applied_revision
+                                       && !present_in_table;
+
+            if (topic_removed && !is_controller) {
+                vlog(
+                  clusterlog.trace,
+                  "can't update leadership of the removed topic {}",
+                  topic.tp_ns);
+                break;
+            }
+            do_update_partition_leader(
+              is_controller, t_it, p.id, p.revision_id, p.term, p.leader_id);
+            co_await ss::coroutine::maybe_yield();
+            if (topic_map_version_snapshot != _topic_map_version) {
+                auto result = _topic_leaders.try_emplace(topic.tp_ns);
+                t_it = result.first;
+                present_in_table = _topic_table.local().contains(topic.tp_ns);
+                last_applied_revision
+                  = _topic_table.local().last_applied_revision();
+                topic_map_version_snapshot = _topic_map_version;
+            }
+        }
+    }
+}
+
+void partition_leaders_table::do_update_partition_leader(
+  bool is_controller,
+  topics_t::iterator t_it,
+  model::partition_id p_id,
   model::revision_id revision_id,
   model::term_id term,
   std::optional<model::node_id> leader_id) {
-    auto key = leader_key_view{
-      model::topic_namespace_view(ntp), ntp.tp.partition};
-
-    const auto is_controller = ntp == model::controller_ntp;
-    /**
-     * Use revision to differentiate updates for the topic that was
-     * deleted from topics table from the one which are processed before the
-     * topic is known to node local topic table.
-     * Controller is a special case as it is not present in topic table.
-     *
-     */
-    const auto topic_removed
-      = revision_id <= _topic_table.local().last_applied_revision()
-        && !_topic_table.local().contains(model::topic_namespace_view(ntp));
+    auto [p_it, new_entry] = t_it->second.emplace(
+      p_id,
+      leader_meta{
+        .current_leader = leader_id,
+        .update_term = term,
+        .partition_revision = revision_id});
 
-    if (topic_removed && !is_controller) {
-        vlog(
-          clusterlog.trace,
-          "can't update leadership of the removed topic {}",
-          ntp);
-        return;
-    }
+    const model::ntp ntp(t_it->first.ns, t_it->first.tp, p_id);
```

Comment:
This potentially voids the benefits that we are getting as it likely causes an alloc + a memcpy.

It seems to be used:
 - In the trace logs: Can we just was pass in a ref to the ns_tp and then use that plus the p_id in the log statements?
 - Later down in the call to `_watchers.notify`: Can we just construct it then or work around somehow else (passing ntp twice there seems weird). I expect that if statement to be rare anyway.